apis/v1alpha1/validation.go

```diff
@@ -33,11 +35,64 @@ func ValidateResourceMapping(m *ResourceMapping) (vErr error) {
 			vErr = errors.Join(vErr, fmt.Errorf("invalid owner: %w", err))
 		}
 	}
-	if m.Resource.Provider == "" {
-		vErr = errors.Join(vErr, fmt.Errorf("empty resource provider"))
-	}
+
 	if _, ok := m.Annotations.AsMap()[AnnotationKeyAssetInfo]; ok {
 		vErr = errors.Join(vErr, fmt.Errorf("reserved key is included: %s", AnnotationKeyAssetInfo))
 	}
+
+	if err := validateResource(m.Resource); err != nil {
+		vErr = errors.Join(vErr, fmt.Errorf("invalid Resourc: %w", err))
+	}
+
 	return
 }
+
+func validateResource(r *Resource) (vErr error) {
+	if r.Name == "" {
+		vErr = errors.Join(vErr, fmt.Errorf("empty resource name"))
+	}
+
+	if r.Provider == "" {
+		vErr = errors.Join(vErr, fmt.Errorf("empty resource provider"))
+	}
+
+	if err := validateAndNormalizeSubscope(r); err != nil {
+		vErr = errors.Join(vErr, fmt.Errorf("invalid subscope: %w", err))
+	}
+
+	return
+}
+
+func validateAndNormalizeSubscope(r *Resource) error {
+	if r.Subscope == "" {
+		return nil
+	}
+
+	u, err := url.Parse(r.Subscope)
+	if err != nil {
+		return fmt.Errorf("failed to parse subscope string: %w", err)
+	}
+
+	_, err = url.ParseQuery(u.RawQuery)
+	if err != nil {
+		return fmt.Errorf("failed to parse qualifier string: %w", err)
+	}
+
+	// check if key value pairs are in alphabetical order
+	// and if key is in lower case
+	s := strings.Split(u.RawQuery, "&")
+
+	for i := 0; i < len(s); i++ {
```

Comment:
This part is difficult to understand. Given a example, subscope is "parent/foo/child/bar?key2=value2&key1=value1", what will be the value of u.RawQuery?